src/libcore/pin.rs

```diff
@@ -191,15 +297,46 @@ impl<P: Deref> Pin<P> {
     /// not guarantee that the data `P` points to is pinned, constructing a
     /// `Pin<P>` is undefined behavior.
     ///
+    /// By using this method, you are making a promise about the `P::Deref` and
+    /// `P::DerefMut` implementations, if they exist. Most importantly, they
+    /// must not move out of their `self` arguments: `Pin::as_mut` and `Pin::as_ref`
+    /// will call `DerefMut::deref_mut` and `Deref::deref` *on the pinned pointer*
+    /// and expect these methods to uphold the pinning invariants.
+    /// Moreover, by calling this method you promise that the reference `P`
+    /// dereferences to will not be moved out of again; in particular, it
+    /// must not be possible to obtain a `&mut P::Target` and then
+    /// move out of that reference (using, for example [`replace`]).
```

Comment:
This seems a good candidate for hoisting to the module-level docs.

Reply:
I have added something along these lines to the module-level docs, please have a look.